File: src/main/java/io/dddrive/util/Formatter.java

package io.dddrive.util;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.text.NumberFormat;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

public class Formatter {

	public static final Formatter INSTANCE = new Formatter();

	private final DecimalFormat decimalFormatter;
	private final DateTimeFormatter dateFormatter;
	private final DateTimeFormatter isoDateFormatter;

	public Formatter() {
		this.decimalFormatter = (DecimalFormat) NumberFormat.getInstance(Locale.GERMAN);
		DecimalFormatSymbols symbols = this.decimalFormatter.getDecimalFormatSymbols();
		symbols.setGroupingSeparator('\'');
		this.decimalFormatter.setDecimalFormatSymbols(symbols);
		this.dateFormatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
		this.isoDateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
	}

	public String formatNumber(Object value) {
		return this.decimalFormatter.format(value);
	}

	public String formatMonetaryValue(Object value, String currency) {
		return this.decimalFormatter.format(value) + " " + currency;
	}

	public String formatValueWithUnit(Object value, String unit) {
		return this.decimalFormatter.format(value) + " " + unit;
	}

	public String formatDate(LocalDate value) {
		return this.dateFormatter.format(value);
	}

	public String formatIsoDate(LocalDate value) {
		return this.isoDateFormatter.format(value);
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/util/Formatter.kt
package io.zeitwert.fm.util

import java.text.DecimalFormat
import java.text.NumberFormat
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.*

object Formatter {

	private val decimalFormatter: DecimalFormat
	private val dateFormatter: DateTimeFormatter
	private val isoDateFormatter: DateTimeFormatter

	init {
		this.decimalFormatter = NumberFormat.getInstance(Locale.GERMAN) as DecimalFormat
		val symbols = this.decimalFormatter.decimalFormatSymbols
		symbols.setGroupingSeparator('\'')
		this.decimalFormatter.decimalFormatSymbols = symbols
		this.dateFormatter = DateTimeFormatter.ofPattern("dd.MM.yyyy")
		this.isoDateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
	}

	fun formatNumber(value: Any?): String = this.decimalFormatter.format(value)

	fun formatMonetaryValue(
		value: Any?,
		currency: String?,
	): String = this.decimalFormatter.format(value) + " " + currency

	fun formatValueWithUnit(
		value: Any?,
		unit: String?,
	): String = this.decimalFormatter.format(value) + " " + unit

	fun formatDate(value: LocalDate): String = this.dateFormatter.format(value)

	fun formatIsoDate(value: LocalDate): String = this.isoDateFormatter.format(value)

}
